### PR TITLE
[reward] feat: add HTTP reward service adapter for external scorer services

### PR DESCRIPTION
**Pull request by ruihanglix**
### What does this PR do?

Adds support for using external HTTP reward services (e.g., PaddleOCR-based scorer) as reward functions in RL training. This enables developers to deploy custom reward models as standalone HTTP services and connect them to verl-omni without modifying any framework code.

Key changes:
1. **`MultiVisualRewardManager` now passes extra config fields to reward functions** — any field beyond `path`/`name`/`weight` in `reward.reward_functions.<key>.*` is forwarded to the corresponding `compute_score` function via kwargs filtering. This is a generic mechanism that enables per-reward custom configuration (e.g., `server_url`, `timeout`).
2. **New generic HTTP scorer client** (`http_scorer_client.py`) — an async `compute_score` adapter that sends images to any HTTP scorer service using the pickle-based protocol.
3. **New data preprocessing script** (`qwenimage_ocr_http_service.py`) — prepares OCR data in a format compatible with the HTTP scorer service (full prompt as ground_truth instead of bare text).
4. **Example training script** (`run_qwen_image_ocr_reward_server.sh`) — demonstrates training with an external OCR reward server, no internal reward model required.

### Motivation

This addresses the RFC discussion in #103 about supporting custom reward models. The implementation follows **Option A (Custom Reward Function via Adapter)** with several modifications purposed by me:
- Users deploy their own HTTP reward server externally
- A `compute_score` adapter calls the endpoint
- The adapter plugs into the existing `MultiVisualRewardManager` path with no framework changes beyond the extra_args forwarding

This approach:
- Requires **zero GPU resources** for reward computation in the training cluster (reward runs on a separate service)
- Supports **any reward model backend** (PaddleOCR, HPSv3, aesthetic models, etc.) as long as it exposes the pickle-based HTTP interface
- Is **fully concurrent** — since `compute_score` is async and `RewardLoopWorker.compute_score_batch` uses `asyncio.gather`, all samples in a batch hit the server concurrently

### Checklist Before Starting

- [x] Search for similar PRs: https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+reward+service
- [x] Format the PR title as `[{modules}] {type}: {description}`

### Test

Validated on 4x NVIDIA B300 GPUs with the Qwen-Image OCR FlowGRPO recipe using an external PaddleOCR reward server (separate GPU).

Training ran for 180+ steps with `batch_size=32`, `n=16`, `lr=3e-5`, `image_resolution=512x512`, `test_freq=10`.

**Results (training reward — `critic/rewards/mean`):**

| Step | 10 | 30 | 50 | 70 | 100 | 130 | 150 | 180 |
|------|----|----|----|----|-----|-----|-----|-----|
| reward | 0.06 | 0.10 | 0.31 | 0.54 | 0.62 | 0.77 | 0.79 | 0.67 |

**Validation reward (`val-core/flow_grpo/ocr/reward/mean@1`):**

| Step | 10 | 30 | 50 | 70 | 100 | 130 | 150 | 180 |
|------|----|----|----|----|-----|-----|-----|-----|
| reward | 0.31 | 0.42 | 0.66 | 0.77 | 0.81 | 0.89 | 0.86 | 0.81 |

Peak validation reward: **0.886** at step 130.

**Training curve (`critic/rewards/mean`):**

image

**Validation curve (`val-core/flow_grpo/ocr/reward/mean@1`):**

image

The model rapidly improves OCR accuracy from ~0.06 to ~0.8 within 100 steps and maintains high performance. The slight drop after step 150 is expected variance from the stochastic reward evaluation.

**Qualitative examples (validation generations):**

Prompt 1: *"A wooden trail marker stands in a dense forest, its surface weathered by time. The words **"Hidden Trail"** are carved deeply into the wood, surrounded by moss and foliage."*

| Step 10 (score: 0.0) | Step 30 (score: 0.0) | Step 50 (score: 0.0) | Step 70 (score: 1.0) | Step 100 (score: 1.0) |
|---|---|---|---|---|
| image | image | image | image | image |

Prompt 2: *"A birthday card interior with a whimsical, pastel-colored design, featuring the message **"Make A Wish"** in elegant, cursive handwriting, surrounded by sparkling candles and colorful confetti."*

| Step 10 (score: 0.33) | Step 30 (score: 0.67) | Step 50 (score: 1.0) | Step 70 (score: 0.89) | Step 100 (score: 1.0) |
|---|---|---|---|---|
| image | image | image | image | image |

Key observations:
- The model learns to render accurate text by step 50-100, going from illegible or missing text to perfect OCR scores.
- Image composition and scene detail remain high quality throughout — the reward service focuses purely on text accuracy without degrading visual quality.
- The HTTP reward server introduces negligible latency: `timing_s/reward` is essentially 0, confirming that async concurrent calls to the server are efficient.

### API and Usage Example

```bash
# 1. Start external OCR reward server (separate process/machine)
# See: https://github.com/deepgenteam/deepgen_rl/tree/main/rewards_services/api_services/ocr_scorer_service
# The service interface format follows: https://github.com/yifan123/flow_grpo#3-reward-preparation
cd rewards_services/api_services/ocr_scorer_service
bash run.sh  # Starts on port 19082

# 2. Prepare data (stores full prompt as ground_truth for HTTP service)
python examples/flowgrpo_trainer/data_process/qwenimage_ocr_http_service.py \
    --input_dir ~/dataset/ocr/ --output_dir ~/data/ocr_http

# 3. Run training
OCR_REWARD_SERVER_URL=http://<server-ip>:19082 \
    bash examples/flowgrpo_trainer/run_qwen_image_ocr_reward_server.sh
```

The key config pattern for HTTP reward services:

```bash
reward.reward_model.enable=False \
reward.reward_manager.name=MultiVisualRewardManager \
"+reward.reward_functions.ocr.path=pkg://verl_omni.utils.reward_score.http_scorer_client" \
'+reward.reward_functions.ocr.name=compute_score' \
'+reward.reward_functions.ocr.weight=1.0' \
"+reward.reward_functions.ocr.server_url=http://localhost:19082"
```

The `server_url` field is passed directly to `compute_score` via the new extra_args mechanism in `MultiVisualRewardManager`. Any custom field can be added this way — the framework uses signature-based filtering to pass only declared parameters.

### Reward Service Interface

The HTTP scorer client (`http_scorer_client.py`) communicates with external reward services using a pickle-based protocol, following the interface defined in [flow_grpo](https://github.com/yifan123/flow_grpo#3-reward-preparation). A reference implementation is available at [deepgen_rl/ocr_scorer_service](https://github.com/deepgenteam/deepgen_rl/tree/main/rewards_services/api_services/ocr_scorer_service).

**Request** (POST, body is `pickle.dumps(payload)`):

```python
{
    "images": List[bytes],    # JPEG-encoded image bytes
    "prompts": List[str],     # prompt/ground_truth strings
    "metadata": dict          # optional metadata (currently empty)
}
```

**Response** (body is `pickle.dumps(response)`):

```python
{
    "scores": List[float]     # one score per image, typically in [0, 1]
}
```

Any service that implements this interface can be used as a reward function — PaddleOCR, HPSv3, aesthetic scorers, CLIP-based scorers, etc. The service runs independently and can use any framework (PaddlePaddle, PyTorch, ONNX, etc.) without conflicting with the training environment.

#### High-level architecture

```
┌──────────────────────────────────────────────────────────────────┐
│               External HTTP Reward Server                         │
│  (e.g., PaddleOCR scorer on separate GPU/machine)                │
│  POST / → pickle({images, prompts, metadata})                    │
│  Response → pickle({scores: List[float]})                        │
└──────────────────────────────────┬───────────────────────────────┘
                                   │ HTTP (async, concurrent)
                                   │
┌──────────────────────────────────┼───────────────────────────────┐
│  MultiVisualRewardManager        │                               │
│                                  ▼                               │
│  run_single(data):                                               │
│    all_kwargs = {solution_image, ground_truth, ...}              │
│    sub_kwargs = {**all_kwargs, **extra_args}  ← NEW: server_url  │
│    filtered_kwargs = _filter_kwargs(sub_kwargs, sig)             │
│    score = await compute_score(**filtered_kwargs)                │
└──────────────────────────────────────────────────────────────────┘
                                   │
┌──────────────────────────────────┼───────────────────────────────┐
│  http_scorer_client.compute_score(solution_image, ground_truth,  │
│                                   server_url, **kwargs)          │
│    1. tensor → PIL → JPEG bytes                                  │
│    2. pickle.dumps({images: [bytes], prompts: [gt], metadata})   │
│    3. aiohttp.post(server_url, data=payload)                     │
│    4. pickle.loads(response) → {"score": float}                  │
└──────────────────────────────────────────────────────────────────┘
```

#### Design decisions

1. **Extra args via config (not env vars):** Each reward function receives its own config fields (e.g., `server_url`) through the `reward.reward_functions.<key>.*` namespace. This is more declarative than env vars and naturally supports multiple HTTP reward services with different endpoints.

2. **Generic client, not OCR-specific:** `http_scorer_client.py` works with any scorer service that implements the pickle-based protocol (`{images, prompts, metadata} → {scores}`). The same client can be reused for HPS, aesthetic, or any future HTTP scorer.

3. **No framework plumbing needed:** By extending `MultiVisualRewardManager` to forward extra config fields, we avoid adding new config schemas or wiring code. The change is 6 lines in `multi.py`.

4. **Async by design:** The `compute_score` function is `async` using `aiohttp`, which means all samples in a batch hit the server concurrently via `asyncio.gather` in `RewardLoopWorker.compute_score_batch`. No serial bottleneck.

5. **Separate data preprocessing:** The HTTP scorer service expects the full prompt (with quoted text) as ground_truth, while the GenRM-based scorer expects bare text. A separate data preprocessing script (`qwenimage_ocr_http_service.py`) handles this difference cleanly.

#### File-by-file changes

| File | Change |
|------|--------|
| `verl_omni/reward_loop/reward_manager/multi.py` | Collect extra config fields per reward function entry and merge them into kwargs before filtering. (+9 lines) |
| `verl_omni/utils/reward_score/http_scorer_client.py` | New. Generic async HTTP scorer client implementing the `compute_score` interface. |
| `examples/flowgrpo_trainer/data_process/qwenimage_ocr_http_service.py` | New. Data preprocessing for HTTP scorer service format. |
| `examples/flowgrpo_trainer/run_qwen_image_ocr_reward_server.sh` | New. Example training script using external OCR reward server. |

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [x] Apply pre-commit checks.
- [x] Backward compatible — existing configs work unchanged (extra_args defaults to empty dict).
- [x] Validated with end-to-end training experiment (180+ steps, 4xB300).
- [x] No internal reward model GPU resources required for this configuration.

### CODE REVIEW

File: verl_omni/utils/reward_score/http_scorer_client.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        {
+            "images": [image_bytes],
+            "prompts": [ground_truth],
+            "metadata": {},
+        }
+    )
+
+    timeout = aiohttp.ClientTimeout(total=120)
+    async with aiohttp.ClientSession(timeout=timeout) as session:
+        async with session.post(server_url, data=payload) as resp:
```
[the remark this answers is not in the document]
Comment (ruihanglix) [reply verl_omni/utils/reward_score/http_scorer_client.py]: https://github.com/verl-project/verl-omni/pull/116/commits/2fbad52578eb114f88d295cbfa1a47e821433251

File: verl_omni/utils/reward_score/http_scorer_client.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    return Image.fromarray(image)
+
+
+def _serialize_image(pil_image: Image.Image) -> bytes:
+    """Serialize a PIL image to JPEG bytes."""
+    if pil_image.mode != "RGB":
+        pil_image = pil_image.convert("RGB")
+    buf = io.BytesIO()
+    pil_image.save(buf, format="JPEG")
+    return buf.getvalue()
```
Comment (zhtmike) [inline verl_omni/utils/reward_score/http_scorer_client.py:52]: it is suggested to put this under thread pool.  It is a CPU-heavy operations and may block the event loop of asynchronous reward call.

You can check our current ocr reward implementation

**Comment by zhtmike** [comment]
And better to have a document page of how to use this feature

File: verl_omni/utils/reward_score/http_scorer_client.py (hunk 2 above)
Comment (ruihanglix) [reply verl_omni/utils/reward_score/http_scorer_client.py:52]: https://github.com/verl-project/verl-omni/pull/116/commits/de179e1305ae5f358334d08611d9ca08b5dcaa9c

**Comment by ruihanglix** [comment]
> And better to have a document page of how to use this feature

https://github.com/verl-project/verl-omni/pull/116/commits/de179e1305ae5f358334d08611d9ca08b5dcaa9c, added `docs/start/http_scorer.md`

**Comment by zhtmike** [comment]
LGTM, @SamitHuang do want to take a look?

**Review by zhtmike — APPROVED** [review]

**Review by zhtmike — APPROVED** [review]
Thank you for your PR!
